### PR TITLE
Drop support for EOL Python 3.7

### LINKED ISSUE #448: Drop support for Python 3.7
**Is your feature request related to a problem? Please describe.**
Python 3.7 end-of-life is [scheduled](https://devguide.python.org/versions/#supported-versions) on 2023-06-23. There's no need to support it in new Darker releases after that. Old versions will still be guaranteed to work.

**Describe the solution you'd like**
Drop Python 3.7 from compatible Python versions. Update documentation and build scripts as well. Bump the major version.

**Describe alternatives you've considered**
We could keep supporting old Python versions, but it would encourage users to keep using software which doesn't receive security updates. Also, we couldn't benefit from new language features if we were forced to stay compatible.

**Additional context**
Files to update:
- `setup.cfg`
- `.github/workflows/python-package.yml`
- `.github/workflows/pyupgrade.yml`
- `src/darker/tests/git.py` (`shlex_join`)
- `src/darker/tests/linting.py` (`shlex_join`)
- `src/darker/tests/test_black_diff.py`
- `src/darker/tests/test_command_line.py`
- `src/darker/tests/test_main.py`

This issue can be used as a template in October 2024 when Python 3.8 has its scheduled end-of-life.

### PR DESCRIPTION
**Pull request by hugovk**
Follow on from #537. Fixes #448.

### CODE REVIEW

File: src/darker/git.py [hunk 1]
```diff
@@ -29,19 +29,7 @@
...
-
-        :param split_command: The elements on the command line
-        :return: The command line as one string, with appropriate quoting
-
-        """
-        return " ".join(shlex.quote(arg) for arg in split_command)
-
-else:
-    shlex_join = shlex.join
+shlex_join = shlex.join
```
Comment (akaihola) [inline src/darker/git.py:32]: Could we just use `shlex.join` directly everywhere?

File: src/darker/import_sorting.py [hunk 2]
```diff
@@ -33,6 +26,7 @@ def isort_code(*args: Any, **kwargs: Any) -> str:  # type: ignore[misc]
                 "An incompatible 'isort' package was found. Please install version"
                 " 5.0.0 or later."
             )
+
```
Comment (akaihola) [inline src/darker/import_sorting.py:29]: These formatting changes most probably have nothing to do with dropping Python 3.7 support either. @hugovk was this suggested by a recent Black version, or is there some other reason?

I'd prefer to collect formatting updates in a separate PR.

File: src/darker/black_diff.py [hunk 3]
```diff
@@ -32,11 +32,11 @@
   based on whether reformats touch user-edited lines
 
 """
+
```
Comment (akaihola) [inline src/darker/black_diff.py:35]: This formatting changes most probably has nothing to do with dropping Python 3.7 support either. @hugovk was this suggested by a recent Black version, or is there some other reason?

I'd prefer to collect formatting updates in a separate PR.

File: setup.cfg [hunk 4]
```diff
@@ -94,6 +92,8 @@ ignore =
     D415
     # E231 missing whitespace after ','
     E231
+    # E704 multiple statements on one line (def)
+    E704
```
Comment (akaihola) [inline setup.cfg:96]: These Black related linting changes most probably has nothing to do with dropping Python 3.7 support. @hugovk probably this was added due to running a recent Black version on the Darker codebase, or is there some other reason?

I'd prefer to collect formatting updates in a separate PR.

**Review by akaihola — CHANGES_REQUESTED** [review]
Thanks @hugovk for the PR!

Only two request, as indicated in my review comments
1. Let's separate Darker code base formatting changes into another PR (or drop them altogether).
2. Replace unnecessary `shlex_join` with `shlex.join` everywhere.

I tried reverting all the formatting changes and running `darker --diff -r master .` on the Darker code base. No matter if I used Black 24.1.1 or 24.2.0, Darker doesn't indicate any required formatting changes. So I'd say let's make this PR cleaner by doing away with the formatting changes.

File: src/darker/git.py [hunk 5]
```diff
@@ -319,15 +307,13 @@ def _git_check_output_lines(
 @overload
 def _git_check_output(
     cmd: List[str], cwd: Path, *, exit_on_error: bool = ..., encoding: None = ...
-) -> bytes:
-    ...
+) -> bytes: ...
```
[the remark this answers is not in the document]
Comment (hugovk) [reply src/darker/git.py:310]: Yes, they're from Black's new 2024 style. I saw in the contributing guidelines that Black should be run, so only ran them on the changed files in this PR.

But makes sense to make the new 2024 style changes to another PR, I'll do that too.

File: src/darker/import_sorting.py (hunk 2 above)
Comment (hugovk) [reply src/darker/import_sorting.py:29]: Yes, from Black.

File: src/darker/black_diff.py (hunk 3 above)
Comment (hugovk) [reply src/darker/black_diff.py:35]: Yes, from Black.

File: setup.cfg (hunk 4 above)
Comment (hugovk) [reply setup.cfg:96]: Yes, from Black.

File: src/darker/git.py (hunk 1 above)
Comment (hugovk) [reply src/darker/git.py]: Good idea, updated!

**Comment by hugovk** [comment]
Would you like me to undo the Black changes in this PR, or wait for https://github.com/akaihola/darker/pull/543 to be merged first and then rebase/update this one?

**Comment by akaihola** [comment]
> Would you like me to undo the Black changes in this PR, or wait for #543 to be merged first and then rebase/update this one?

I'd prefer removing those changes from this PR since it doesn't seem to affect the test results.

File: setup.cfg (hunk 4 above)
Comment (hugovk) [inline setup.cfg:96]:
```suggestion
```

File: src/darker/black_diff.py (hunk 3 above)
Comment (hugovk) [inline src/darker/black_diff.py:35]:
```suggestion
```

File: src/darker/git.py [hunk 6]
```diff
@@ -319,15 +306,13 @@ def _git_check_output_lines(
...
```
Comment (hugovk) [inline src/darker/git.py:309]:
```suggestion
) -> bytes:
    ...
```

File: src/darker/git.py [hunk 7]
```diff
@@ -319,15 +306,13 @@ def _git_check_output_lines(
...
 
 
 @overload
 def _git_check_output(
     cmd: List[str], cwd: Path, *, exit_on_error: bool = ..., encoding: str
-) -> str:
-    ...
+) -> str: ...
```
Comment (hugovk) [inline src/darker/git.py:315]:
```suggestion
) -> str:
    ...
```

File: src/darker/import_sorting.py (hunk 2 above)
Comment (hugovk) [inline src/darker/import_sorting.py:29]:
```suggestion
```

File: src/darker/import_sorting.py [hunk 8]
```diff
@@ -52,6 +46,7 @@ def isort_code(*args: Any, **kwargs: Any) -> str:  # type: ignore[misc]
 
 class IsortArgs(TypedDict, total=False):
     """Types for command line arguments for ``isort``"""
+
```
Comment (hugovk) [inline src/darker/import_sorting.py:49]:
```suggestion
```

**Review by hugovk — COMMENTED** [review]
Revert Black 2024 updates

File: src/darker/git.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (akaihola) [reply src/darker/git.py]: > [...]  in the contributing guidelines that Black should be run, so only ran them on the changed files in this PR.

Hm, maybe I'll need to improve the wording – what I originally meant was to only run Black and isort on the modified lines.

**Pull request merged**